casper/src/test/scala/coop/rchain/casper/batch2/BlockValidationLogicSpec.scala

```diff
@@ -0,0 +1,150 @@
+
+      hashInValid shouldBe false
+    }
+  }
+
+  "Field format validation" should "succeed on a valid block and fail on empty fields" in {
+    val (privateKey, publicKey) = Secp256k1.newKeyPair
+    val genesis                 = signedBlock(privateKey, publicKey)
+
+    BlockValidationLogic.formatOfFields(genesis) shouldBe true
+    BlockValidationLogic.formatOfFields(genesis.copy(blockHash = ByteString.EMPTY)) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(sig = ByteString.EMPTY)) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(sigAlgorithm = "")) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(shardId = "")) shouldBe false
+    BlockValidationLogic.formatOfFields(genesis.copy(postStateHash = ByteString.EMPTY)) shouldBe false
+  }
+
+  // TODO: This test doesn't make sense because the signing is done by the algorithm in the ValidatorIdentity.
+  //  The algorithm specified for the block is not used
+  "Block signature validation" should "return false on unknown algorithms" in {
+    val blockUnknownAlg = getRandomBlock().copy(sigAlgorithm = "unknownAlgorithm")
+    val blockRsaAlg     = getRandomBlock().copy(sigAlgorithm = "RSA")
+
+    BlockValidationLogic.blockSignature(blockUnknownAlg) shouldBe false
+    BlockValidationLogic.blockSignature(blockRsaAlg) shouldBe false
+  }
+
+  it should "return false on invalid secp256k1 signatures" in {
+    val (privateKey, publicKey) = Secp256k1.newKeyPair
+    val (_, wrongPk)            = Secp256k1.newKeyPair
+    val empty                   = ByteString.EMPTY
+    val invalidKey              = "abcdef1234567890".unsafeHexToByteString
+
+    val block0 = signedBlock(privateKey, publicKey).copy(sender = empty)
+    val block1 = signedBlock(privateKey, publicKey).copy(sender = invalidKey)
+    val block2 =
+      signedBlock(privateKey, publicKey).copy(sender = ByteString.copyFrom(wrongPk.bytes))
+    val block3 = signedBlock(privateKey, publicKey).copy(sig = empty)
+    val block4 = signedBlock(privateKey, publicKey).copy(sig = invalidKey)
+    val block5 = signedBlock(privateKey, publicKey).copy(sig = block0.sig) // wrong sig
+    val blocks = Vector(block0, block1, block2, block3, block4, block5)
+
+    blocks.exists(BlockValidationLogic.blockSignature) shouldBe false
+  }
+
+  it should "return true on valid secp256k1 signatures" in {
+    val n                       = 6
+    val (privateKey, publicKey) = Secp256k1.newKeyPair
+    val result = (0 until n).foldLeft(true) {
+      case (res, _) =>
+        val block = signedBlock(privateKey, publicKey)
+        BlockValidationLogic.blockSignature(block) && res
+    }
+    result shouldBe true
+  }
+
+  "Future deploy validation" should "work" in {
+    val deploy = createDeploy(-1L)
+    val block  = getRandomBlock(setDeploys = Seq(ProcessedDeploy.empty(deploy)).some)
+    val status = BlockValidationLogic.futureTransaction(block)
+
+    status shouldBe Right(Valid)
+  }
```

Comment:
We want to have generative tests whenever possible. Testing range of numbers are good example for that.

If you want to have fixed values for testing you can use `TableDrivenPropertyChecks` (included in ScalaCheckPropertyChecks). HEre is one example using fixed set of values for IP addresses.
https://github.com/rchain/rchain/blob/ca452d75dd85a65442e4b76b5f8d33cbe5b71e2a/comm/src/test/scala/coop/rchain/comm/rp/HandleProtocolHandshakeSpec.scala#L50-L77

Reply:
[Done](https://github.com/rchain/rchain/pull/3793/commits/d5b040d334e2ae0cc707d3f1ba9829e1527888a9).